File: certification_server/src/services/communication/traffic_station_svc.go
package communication

import (
	"context"
	"strconv"
	"strings"

	authcontrolif "certification_server/src/iface/authcontrol"
	communicationif "certification_server/src/iface/communication"
	authmodel "certification_server/src/models/auth"
	modelsystem "certification_server/src/models/system"
)

var _ communicationif.ITrafficStation = (*TrafficStationService)(nil)

// TrafficStationService 提供认证中心统一入站流量站点实现。
type TrafficStationService struct {
	routingPipeline communicationif.IRoutingPayloadPipeline
	authControl     authcontrolif.IInboundAuthControl
}

func NewTrafficStationService(
	routingPipeline communicationif.IRoutingPayloadPipeline,
	authControl authcontrolif.IInboundAuthControl,
) communicationif.ITrafficStation {
	return &TrafficStationService{
		routingPipeline: routingPipeline,
		authControl:     authControl,
	}
}

func (s *TrafficStationService) HandleInbound(
	ctx context.Context,
	req *communicationif.InboundTrafficRequest,
) (*communicationif.TrafficDecision, error) {
	if req == nil || req.Route == nil {
		return nil, &modelsystem.ErrTrafficInboundRequestInvalid
	}
	if s.routingPipeline == nil {
		return nil, &modelsystem.ErrTrafficStationDependenciesRequired
	}

	policy, err := s.routingPipeline.BuildInboundPolicy(ctx, req.Route)
	if err != nil {
		return nil, err
	}
	if policy == nil || policy.RouteProfile == nil {
		return nil, &modelsystem.ErrRouteProfileNotFound
	}

	metadata := map[string]string{
		"flow_category":       string(policy.RouteProfile.FlowCategory),
		"security_policy":     string(policy.RouteProfile.SecurityPolicy),
		"operation":           strings.TrimSpace(policy.RouteProfile.Operation),
		"target_service_type": strings.TrimSpace(policy.RouteProfile.TargetServiceType),
		"target_service_name": strings.TrimSpace(policy.RouteProfile.TargetServiceName),
		"target_endpoint":     strings.TrimSpace(policy.RouteProfile.TargetEndpoint),
	}
	if len(policy.RequiredScopes) > 0 {
		metadata["required_scopes"] = strings.Join(policy.RequiredScopes, ",")
	}
	for key, value := range policy.Tags {
		trimmed := strings.TrimSpace(value)
		if trimmed == "" {
			continue
		}
		metadata[key] = trimmed
	}

	decision := &communicationif.TrafficDecision{
		Accepted: true,
		Reason:   "accepted",
		Profile:  policy.RouteProfile,
		Metadata: metadata,
	}

	if s.authControl == nil {
		return decision, nil
	}

	controlResult, err := s.authControl.EnforceInbound(
		ctx,
		&authcontrolif.InboundControlRequest{
			RateLimitInput: buildInboundRateLimitInput(req, policy),
		},
	)
	if err != nil {
		return nil, err
	}
	if controlResult == nil || controlResult.RateLimitDecision == nil {
		return nil, &modelsystem.ErrRateLimitRequestInvalid
	}

	decision.Metadata["rate_limit_allowed"] = strconv.FormatBool(controlResult.RateLimitDecision.Allowed)
	decision.Metadata["rate_limit_rule_id"] = strings.TrimSpace(controlResult.RateLimitDecision.ViolatedRuleID)
	decision.Metadata["rate_limit_subject_key"] = strings.TrimSpace(controlResult.RateLimitDecision.SubjectKey)
	decision.Metadata["rate_limit_remaining"] = strconv.FormatInt(controlResult.RateLimitDecision.Remaining, 10)
	decision.Metadata["rate_limit_retry_after_sec"] = strconv.FormatInt(controlResult.RateLimitDecision.RetryAfterSec, 10)

	if !controlResult.RateLimitDecision.Allowed {
		decision.Accepted = false
		reason := strings.TrimSpace(controlResult.RateLimitDecision.Reason)
		if reason == "" {
			reason = "rate limited"
		}
		decision.Reason = reason
	}

	return decision, nil
}

func (s *TrafficStationService) SendOutbound(
	ctx context.Context,
	req *communicationif.OutboundTrafficRequest,
) (*communicationif.TrafficDispatchResult, error) {
	if req == nil || req.Route == nil {
		return nil, &modelsystem.ErrTrafficInboundRequestInvalid
	}
	if s.routingPipeline == nil {
		return nil, &modelsystem.ErrTrafficStationDependenciesRequired
	}

	profile, err := s.routingPipeline.ResolveRouteProfile(ctx, req.Route)
	if err != nil {
		return nil, err
	}
	if profile == nil {
		return nil, &modelsystem.ErrRouteProfileNotFound
	}

	return &communicationif.TrafficDispatchResult{
		Profile: profile,
		Payload: req.Payload,
		Metadata: map[string]string{
			"flow_category":       string(profile.FlowCategory),
			"security_policy":     string(profile.SecurityPolicy),
			"operation":           profile.Operation,
			"target_service_type": strings.TrimSpace(profile.TargetServiceType),
			"target_service_name": strings.TrimSpace(profile.TargetServiceName),
			"target_endpoint":     strings.TrimSpace(profile.TargetEndpoint),
		},
	}, nil
}

func buildInboundRateLimitInput(
	req *communicationif.InboundTrafficRequest,
	policy *communicationif.InboundPolicyPlan,
) *authcontrolif.InboundRateLimitInput {
	if req == nil || req.Route == nil || policy == nil || policy.RouteProfile == nil {
		return nil
	}

	headers := map[string]string{}
	for key, value := range req.Headers {
		trimmedKey := strings.ToLower(strings.TrimSpace(key))
		trimmedValue := strings.TrimSpace(value)
		if trimmedKey == "" || trimmedValue == "" {
			continue
		}
		headers[trimmedKey] = trimmedValue
	}

	tags := make(map[string]string, len(policy.Tags)+len(policy.RouteProfile.Metadata))
	for key, value := range policy.RouteProfile.Metadata {
		trimmedValue := strings.TrimSpace(value)
		if trimmedValue == "" {
			continue
		}
		tags[key] = trimmedValue
	}
	for key, value := range policy.Tags {
		trimmedValue := strings.TrimSpace(value)
		if trimmedValue == "" {
			continue
		}
		tags[key] = trimmedValue
	}
	for key, value := range headers {
		tags["header."+key] = value
	}

	return &authcontrolif.InboundRateLimitInput{
		Scope:         authmodel.RateLimitScopeAuth,
		Transport:     strings.TrimSpace(req.Route.Transport),
		Module:        strings.TrimSpace(policy.RouteProfile.TargetServiceName),
		Action:        strings.TrimSpace(policy.RouteProfile.Operation),
		Route:         strings.TrimSpace(req.Route.RouteKey),
		Method:        strings.TrimSpace(req.Route.Method),
		SourceIP:      headers["x-source-ip"],
		GatewayID:     firstNonEmpty(headers["x-gateway-id"], req.Route.SourceService),
		ClientID:      headers["x-client-id"],
		SourceService: strings.TrimSpace(req.Route.SourceService),
		TargetService: strings.TrimSpace(req.Route.TargetService),
		Headers:       headers,
		Tags:          tags,
	}
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if trimmed := strings.TrimSpace(value); trimmed != "" {
			return trimmed
		}
	}
	return ""
}

File: data_worker/src/services/communication/traffic_station_svc.py
from __future__ import annotations

from src.iface.communication.routing_contract import FlowCategory
from src.iface.communication.routing_payload_pipeline import (
    BuildOutboundPayloadRequest,
    IRoutingPayloadPipeline,
)
from src.iface.communication.traffic_station import (
    InboundTrafficRequest,
    ITrafficStation,
    OutboundTrafficRequest,
    TrafficDecision,
    TrafficDispatchResult,
)


class TrafficStationService(ITrafficStation):
    """data_worker 统一流量站点（L2）。"""

    def __init__(self, *, routing_pipeline: IRoutingPayloadPipeline | None) -> None:
        self._routing_pipeline = routing_pipeline

    async def handle_inbound(self, req: InboundTrafficRequest) -> TrafficDecision:
        if req is None or req.flow is None:
            raise ValueError("inbound traffic request is invalid")
        if self._routing_pipeline is None:
            raise RuntimeError("traffic station dependencies are required")

        profile = await self._routing_pipeline.resolve_route_profile(req.flow)
        accepted = True
        reason = "accepted"

        if _requires_target_endpoint(profile.flow_category):
            if not (profile.target_service_name or "").strip():
                accepted = False
                reason = "route_target_unresolved"
            elif not (profile.target_endpoint or "").strip():
                accepted = False
                reason = "route_endpoint_unresolved"

        metadata = {
            "flow_category": profile.flow_category,
            "security_policy": profile.security_policy,
            "target_service": profile.target_service_name,
            "target_endpoint": profile.target_endpoint,
        }
        return TrafficDecision(
            accepted=accepted,
            reason=reason,
            profile=profile,
            metadata=metadata,
        )

    async def send_outbound(self, req: OutboundTrafficRequest) -> TrafficDispatchResult:
        if req is None or req.flow is None:
            raise ValueError("outbound traffic request is invalid")
        if self._routing_pipeline is None:
            raise RuntimeError("traffic station dependencies are required")

        plan = await self._routing_pipeline.build_outbound_payload(
            BuildOutboundPayloadRequest(
                flow=req.flow,
                plain_payload=req.payload,
                additional_data=dict(req.headers or {}),
            )
        )

        target_endpoint = (plan.route_profile.target_endpoint or "").strip()
        if not target_endpoint and plan.target is not None:
            target_endpoint = (plan.target.endpoint or "").strip()

        return TrafficDispatchResult(
            profile=plan.route_profile,
            target_endpoint=target_endpoint,
            payload=plan.plain_payload,
            metadata={
                "flow_category": plan.route_profile.flow_category,
                "security_policy": plan.route_profile.security_policy,
                "target_service": plan.route_profile.target_service_name,
                "target_endpoint": target_endpoint,
            },
        )


def _requires_target_endpoint(category: FlowCategory) -> bool:
    return category in {
        "bootstrap_call",
        "remote_auth_verify",
        "external_auth_forward",
        "module_token_refresh",
    }
